### DISCUSSION TITLE
Autorun of main.py/boot.py does not work with micropython 1.23.0 on Raspberry Pi Pico -- FIXED

### DISCUSSION THREAD

**Discussion by wbeebe** [micropython/micropython#15292] [RP2040 / Pico] [answered]
## _Environment_
- Version of MicroPython: 1.23 (and 1.22)
- Downloaded the firmware from MicroPython GitHub releases
- Which hardware/board: Regular Raspberry Pi Pico
- Thonny version 4.1.4 used to program device
- Host development system is Linux Mint 21.3

## _Details_
I have a number of Raspberry Pi Pico and Pico W boards purchased from Adafruit. The boards all appear to work properly, except that the Pico boards will not autorun main.py/boot.py on powerup, i.e. when the board is plugged into USB. If I use Thonny (version 4.1.4) and type [Ctrl] d in the shell, then MicroPython soft reboots and main.py works as expected.

_**Note that the exact same code works properly on the Raspberry Pi Pico W, that is main.py autoruns when the board is plugged into USB.**_

## _External Reference_

Blog post containing code listing and photo of setup is here:
[moving from circuit python to micro python, part two](https://arcanesciencelab.wordpress.com/2024/06/16/moving-from-circuit-python-to-micro-python-part-two/)

**Comment by robert-hh** [comment]
Please verify that boot.py and main.py have been uploaded to the board's file system. And if yes, try another terminal emulator instead of Thonny, like mpremote, putty, screen, picocom, depending on you PC's OS.

**Reply by wbeebe** [reply]
> Please verify that boot.py and main.py have been uploaded to the board's file system.

It was.

> ...try another terminal emulator ... like mpremote, putty, screen, picocom...

I already tried mpremote:
```
mpremote fs cp main.py :main.py
```
and
```
mpremote ls
ls :
        5394 SSD1306.py
        3615 main.py`
```
And of course a check on my system:
```
ls -l
-rw-rw-r-- 1 mint mint 3615 Jun 17 10:45 main.py
-rw-rw-r-- 1 mint mint 5394 Jun 15 19:42 SSD1306.py
```
I did all that before I posted this. Even with mpremote copying main.py onto the Pico, it still won't autorun when the Pico is plugged into USB for power.

**Reply by robert-hh** [reply]
If main.py is present on the board it will be executed, unless running boot.py flags an error. What do you see is you just us mpremote as terminal. In a bash window, call `mpremote`. You can as well push Ctrl-C to stop a running code and Ctrl-D for soft reset.

**Reply by wbeebe** [reply]
There is no boot.py, just the two files listed in my last reply. This is what happens when I connect via `mpremote`:

```
 ~ mpremote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] to exit this shell

>>> ( Ctrl D typed here)
MPY: soft reboot

 MEM FREE: 224,384 BYTES
 FS TOTAL: 1,441,792 BYTES
 FS  FREE: 1,421,312 BYTES
 PLATFORM: MicroPython-1.23.0-arm--with-newlib4.3.0
      UID: E660C062130D5936
     SSID: RP2-5936
 CPU FREQ: 125,000,000 Hz
      I2C: SoftI2C(scl=27, sda=26, freq=500000)
      I2C: DEVICES FOUND: ['0x23', '0x3d']
      I2C: SSD1306 OLED

  NETWORK: NO WIFI ON DEVICE

MicroPython v1.23.0 on 2024-06-02; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>> 
```

**Reply by robert-hh** [reply]
So output must be done by main.py. It's not created by the standard MicroPython firmware.

**Reply by robert-hh** [reply]
What do you get when you type `help("modules")`?

**Reply by wbeebe** [reply]
I'm not sure what you're driving at, but I do know this:

1. The exact same code executes properly on power-on, on a Raspberry Pi Pico W.
2. The exact same code _does not execute properly_ on power-on, on a Raspberry Pi Pico.

And it's not just this code. I have another main.py file that does nothing but flash the on-board LED. On power-on it fails to start on the regular Raspberry Pi Pico, but works as expected on the Pico W.

Know also that I have three regular Picos exhibiting this failure, and two Pico Ws that are working correctly. All five boards are running "standard MicroPython firmware" downloaded from the MicroPython website.

**Reply by wbeebe** [reply]
To answer your last question:
```
>>> help('modules')
__main__          asyncio/lock      gc                random
_asyncio          asyncio/stream    hashlib           re
_boot             binascii          heapq             rp2
_boot_fat         builtins          io                select
_onewire          cmath             json              struct
_rp2              collections       machine           sys
_thread           cryptolib         math              time
array             deflate           micropython       uasyncio
asyncio/__init__  dht               neopixel          uctypes
asyncio/core      ds18x20           onewire           vfs
asyncio/event     errno             os
asyncio/funcs     framebuf          platform
Plus any modules on the filesystem
>>>
```
I think I'm done here.

**Reply by robert-hh** [reply]
I want to be sure that there is no frozen main.py at your board. That would show up with help("modules"). You told that _that the Pico boards will not autorun main.py/boot.py on powerup_ . But there is something executes at boot resp soft reset.

**Reply by robert-hh** [reply]
So it seems that your main.py expects to run on a Pico W and stops if it's not. can you share your main.py code?

**Comment by sosi-deadeye** [comment]
I used the official [RPI_PICO-20240602-v1.23.0.uf2](https://micropython.org/resources/firmware/RPI_PICO-20240602-v1.23.0.uf2) with RP Pico (no Wifi).

Add `print("Hello from boot")` to your `boot.py` and `print("Hello from main")` to `main.py`. You can do it also with Thonny. Then save both files. Then press `CTRL+D`, which sends an EOF to the board, which resets the board and Thonny does not interfere. It should print both messages to the REPL.

**Comment by bikeNomad** [comment]
Another way to test this is to force a full reset from the REPL when connected (using `machine.reset()`).  This should run code the same way as a power-on reset.

It's possible that your code is actually running at power-up but `mpremote` (or other terminal emulator) and your OS isn't fast enough at making the serial connection to see the result.

To test this theory, add a `time.sleep(10)` at the start of main.py and see if the result is any different (note that `mpremote` will usually do a soft-reset, so instead use
```
mpremote resume
```

to see what's happening without the soft-reset.

**Reply by wbeebe** [reply]
No. I have a link at the top of this thread that leads to the post I wrote about both the code and what I was doing with the code. While you should go read it, here's the photo of the setup I have been testing with.

![53796014944_23cc732499_c](https://github.com/micropython/micropython/assets/4269487/e0d76071-02aa-4186-9e5a-92f5fa5135d3)

That OLED display will display that text and graphics when main.py _**automatically**_ runs. It works just fine with the Pico W, which is what you see at the top of the photo. If I put any of the three regular Picos I currently own on that Treedix breakout board, then it won't run on power-up. If I connect `Thonny` or `mpremote` to the regular Pico (no wireless) port (usually `/dev/ttyACM0`) and then hit [Ctrl] D, causing a soft reboot, then it executes and the display shows what you see in the photo.

This is my last response. I've already forked the MicroPython repo into my local account which I keep up to date with the original, and I have a git clone of that copy on my system. I build it with the ESP-IDF 5.2.2 tool chain for my Espressif ESP32-S3 boards. It looks like I'm going to have to install the ARM Cortex-M0+ toolchain for the RP2040 and follow a very similar path to fix this.

**Reply by robert-hh** [reply]
Looking at the code of main.py it's obvious that opposite to your initial statement main.py runs. 
When getting the output of main.py shown above, did you put the RPi Pico at the same breakout board than the Pico W, with the same display? It's then indeed surprising that the display does not show anything. Not even a random pattern at power-up?
Another question: the main.py print shows two I2C devices on the bus, with 0x3d probably being the SSD1306. What is the other device? Should there be one?

**Reply by robert-hh** [reply]
Just two additional notes: 
- When set at 500kHz, the SCL clock is indeed close to 500 kHz for both Pico W (505kHz) and Pico (511 kHz).
- The SSD1306 is rated at maximum SCL frequency of 400 kHz. It may not matter, but it's out of spec.

**Reply by wbeebe** [reply]
Changing the frequency to a lower value doesn't matter. I set it to 250 kHz (the lowest it can be set) and it still won't execute on power-on.

**Reply by scruss** [reply]
what ssd1306 library did you use? Your code doesn't work with the official library. See my comment below

**Comment by bikeNomad** [comment]
@wbeebe I just tried this version of your `main.py` code (which starts with a 5-second delay to allow communications re-connection) on a Raspberry Pi Pico (_not_ a Pico W), and I saw correct power-on behavior (I was using `minicom`, which re-tries opening the COM port). I did move the imports to the top of the code and added a `print` statement to show the reset cause (1 or 3). This is what I got:

```py
import machine as ma
import binascii
import platform
import os
import gc
import time

print("starting main.py")
time.sleep(5)
 
print(f" reset reason: {ma.reset_cause()}")
print(f" MEM FREE: {gc.mem_free():,} BYTES")
 
UNAME = os.uname().sysname.upper()
stat_vfs = os.statvfs('/')
print(f" FS TOTAL: {stat_vfs[0] * stat_vfs[2]:,} BYTES")
print(f" FS  FREE: {stat_vfs[0] * stat_vfs[3]:,} BYTES")
 
print(f" PLATFORM: {platform.platform()}")
 
UNIQUE_ID = binascii.hexlify(ma.unique_id()).decode('ascii').upper()
print(f"      UID: {UNIQUE_ID}")
SSID = UNAME + '-' + UNIQUE_ID[-4:]
print(f"     SSID: {SSID}")
print(f" CPU FREQ: {ma.freq():,} Hz")
 
# Scan I2C bus for devices
#
# I2C pins for Raspberry Pi Pico W, device I2C1
#
SDA_PIN = 26
SCL_PIN = 27
SOFT_I2C = ma.SoftI2C(scl=ma.Pin(SCL_PIN), sda=ma.Pin(SDA_PIN))
print(f"      I2C: {SOFT_I2C}")
i2c_scanned = SOFT_I2C.scan()
 
if len(i2c_scanned) == 0:
    print("      I2C: No Devices Found")
else:
    print("      I2C: DEVICES FOUND:", [hex(device_address)
        for device_address in i2c_scanned])
 
    # Check if there is an SSD1306 display attached.
    #
    import SSD1306
    import framebuf
 
    if SSD1306.OLED_ADDR in i2c_scanned:
        print("      I2C: SSD1306 OLED")
        #
        # Create instance of SSD1306 class to control the display.
        # Initialize it by clearing everything.
        #
        display = SSD1306.SSD1306_I2C(SOFT_I2C)
        display.fill(0)
        #
        # Create a graphic of the Raspberry Pi logo.
        # Display it twice, one logo for each RP2040 core, similar to what
        # the regular Raspberry Pi does on initial boot.
        # I copied the bytearray for the logo from Raspberry Pi itself:
        # https://github.com/raspberrypi/pico-micropython-examples/tree/master/i2c
        #
        buffer = bytearray(b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x7C\x3F\x00\x01\x86\x40\x80\x01\x01\x80\x80\x01\x11\x88\x80\x01\x05\xa0\x80\x00\x83\xc1\x00\x00C\xe3\x00\x00\x7e\xfc\x00\x00\x4c\x27\x00\x00\x9c\x11\x00\x00\xbf\xfd\x00\x00\xe1\x87\x00\x01\xc1\x83\x80\x02A\x82@\x02A\x82@\x02\xc1\xc2@\x02\xf6>\xc0\x01\xfc=\x80\x01\x18\x18\x80\x01\x88\x10\x80\x00\x8c!\x00\x00\x87\xf1\x00\x00\x7f\xf6\x00\x008\x1c\x00\x00\x0c\x20\x00\x00\x03\xc0\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00")
        raspberry_pi_logo = framebuf.FrameBuffer(buffer, 32, 32, framebuf.MONO_HLSB)
        display.framebuf.blit(raspberry_pi_logo, 0, 33)
        display.framebuf.blit(raspberry_pi_logo, 33, 33)
        #
        # Display the official MicroPython logo
        #
        display.framebuf.fill_rect(0, 0, 32, 32, 1)
        display.framebuf.fill_rect(2, 2, 28, 28, 0)
        display.framebuf.vline(9, 8, 22, 1)
        display.framebuf.vline(16, 2, 22, 1)
        display.framebuf.vline(23, 8, 22, 1)
        display.framebuf.fill_rect(26, 24, 2, 4, 1)
        #
        # Print some identifying text with the graphics, such as version and
        # the identifying string of the Raspberry Pi Pico.
        #
        display.text('MicroPython', 40, 0, 1)
        display.text('-'.join(platform.platform().split('-')[1:3]), 40, 12, 1)
        display.text(SSID, 40, 24, 1)
        display.show()
 
print()
 
try:
    import network
    wifi = network.WLAN(network.STA_IF)
    wifi.active(True)
    access_points = wifi.scan()
    networks = {}
 
    for network in access_points:
        if len(network[0]) > 0 and bytearray(network[0])[0] != 0:
            ssid = network[0].decode('utf-8')
            networks[ssid] = network[3]
 
    for ssid in sorted(networks.keys()):
        print(f"ssid: {ssid:24} rssi: {networks[ssid]}")
except:
    print("  NETWORK: NO WIFI ON DEVICE")
 
print()
```

I started the board up and hit ctrl-D, and got this:
```
MPY: soft reboot
starting main.py
 reset reason: 1
 MEM FREE: 224,064 BYTES
 FS TOTAL: 1,441,792 BYTES
 FS  FREE: 1,429,504 BYTES
 PLATFORM: MicroPython-1.23.0-arm--with-newlib4.3.0
      UID: E660B4400729482A
     SSID: RP2-482A
 CPU FREQ: 125,000,000 Hz
      I2C: SoftI2C(scl=27, sda=26, freq=500000)
      I2C: No Devices Found

  NETWORK: NO WIFI ON DEVICE

MicroPython v1.23.0 on 2024-06-02; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>>
```
Then I turned the power off and back on and got this:

```
reset reason: 1
 MEM FREE: 224,064 BYTES
 FS TOTAL: 1,441,792 BYTES
 FS  FREE: 1,429,504 BYTES
 PLATFORM: MicroPython-1.23.0-arm--with-newlib4.3.0
      UID: E660B4400729482A
     SSID: RP2-482A
 CPU FREQ: 125,000,000 Hz
      I2C: SoftI2C(scl=27, sda=26, freq=500000)
      I2C: No Devices Found

  NETWORK: NO WIFI ON DEVICE

MicroPython v1.23.0 on 2024-06-02; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>>
```

Reset reason 1 is machine.PWRON_RESET, and 3 is machine.WDT_RESET (which is used for soft reset on the RP2).

So it's definitely not that `main.py` is not being executed from power-up on the RPi Pico.

**Comment by scruss** [comment]
@wbeebe - I tried your code, this time on a Pico W with no I2C LED attached. This is the output I get from main.py:
```

 MEM FREE: 182,528 BYTES
 FS TOTAL: 868,352 BYTES
 FS  FREE: 819,200 BYTES
 PLATFORM: MicroPython-1.23.0-arm--with-newlib4.3.0
      UID: E6614C775B2E3535
     SSID: RP2-3535
 CPU FREQ: 125,000,000 Hz
      I2C: SoftI2C(scl=27, sda=26, freq=500000)
      I2C: No Devices Found

ssid: Ruel34                   rssi: -65
ssid: zIEX2LQE                 rssi: -75

MicroPython v1.23.0 on 2024-06-02; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> 

```
In order to see if the code was being run on USB connect with no terminal present, I added these lines at the end:
```python3
import time
led=ma.Pin("LED", ma.Pin.OUT)
led.value(1)
time.sleep(1)
led.value(0)
```
This flashes the internal LED for one second on successful completion of `main.py`. And sure enough I get the LED flash on a Pico W.

To repeat @bikeNomad's result, I tried a regular Raspberry Pi Pico and got the expected result, complete with LED flash.

So I'm very much afraid to say I can't repeat your issue.

One detail I did notice when I tried to add an I²C SSD1306 OLED was that the SSD1306 module you used has some incompatible methods compared to the official [ssd1306](https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/display/ssd1306) module. I installed the official module using `mpremote a1 mip install ssd1306`. So I never got that part to work.

**Comment by wbeebe** [comment, marked as answer]
# SOLUTION
The solution is in two parts.
1. The pins I originally chose for the I2C bus (26 and 27) appear to work for the Pico W, but not for the original Pico. By going back and looking at section 3.6, **I2C** (page 15) of the Raspberry Pi Pico Python SDK document, the examples they give use pins 8 (SDA) and 9 (SCL) or 6 (SDA) and 7 (SCL). By moving the SDA and SCL lines over to those two pins, autorun on power-up occurred with both the Pico and Pico W.
2. The example code in the same section noted above had the frequency at 100 kHz. When the lines were moved and the frequency set to 100 kHz then the file `main.py` executed completely every time and both text and graphics were drawn on the attached OLED display.

The code section that sets up the I2C buss now looks like this:
``` python
SDA_PIN = 8 # Blue wire
SCL_PIN = 9 # Yellow wire
SOFT_I2C = ma.SoftI2C(scl=ma.Pin(SCL_PIN), sda=ma.Pin(SDA_PIN), freq=100000)
```
### UPDATE 18 June 2024
One other issue I've discovered is timing. Based on several comments, I added the following at the start of the `main.py`:
``` python
import time
time.sleep_ms(500)
```
That half-a-second doesn't help the Pico W, but it appears to help the original Pico. With that delay at the start the original Pico starts **_consistently_** every time power is applied. I have to wonder if that half-a-second is allowing something within the RP2040 to settle after power-on.

**Reply by robert-hh** [reply]
Glad you found a solution. Still some details of the resolution are interesting. 
- First, with SoftI2C any GPIO pin pair can be used. And in my test yesterday, I hooked up a I2C device to pins 26 and 27 and it worked using SoftI2C as expected. You could indeed as well use hard I2C at these pins.
- Second, the need to reduce the clock frequency to 100kHz indicates some kind of wiring dependency. It may happen that scan (== 1 byte transfer) works, but longer messages fail. I looked up the schematics of the display and it shows 10k pull-up resistors. With the short wires at your picture that should work at higher frequencies, and I2C on RP2 drops the frequency when the time constant of the rising slope at a pin is too large. But one had to hook up a logic analyzer or oscilloscope to analyze the actual signals. It may be that with 100kHz it works as well at pins 26 and 27.

**Reply by wbeebe** [reply]
No, it did not work on pins 26 and 27. I tested 26 and 27 with the 100 kHz frequency and it still failed with the original Pico. Only pins 8 and 9 (or 6 and 7) worked consistently across both Pico and Pico W. I'm beginning to wonder if there's a hardware issue with regards to those two pins. After all, as I noted above, only the first two pair of I2C pins are shown in the Raspberry Pi Pico Python SDK documentation examples.

**Reply by scruss** [reply]
The [official pinout document for the Raspberry Pi Pico / Pico W](https://datasheets.raspberrypi.com/pico/Pico-R3-A4-Pinout.pdf) has many pin pairs that can be used with hardware I2C. Table 2 (p.16 of the [Raspberry Pi Pico Python SDK](https://datasheets.raspberrypi.com/pico/raspberry-pi-pico-python-sdk.pdf)) shows the default I2C pins for the Raspberry Pi Pico/Pico W boards. But they're not the default I2C pins on every RP2040 board. On my Sparkfun Thing+ RP2040, they're on pins 16 and 17:
```
>>> import machine
>>> i0=machine.I2C(0)
>>> i0
I2C(0, freq=399361, scl=17, sda=16, timeout=50000)
```
